[jiaxinnns] iP

jiaxinnns opened the pull request:
# ReoPro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

ReoPro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

jiaxinnns committed:
Give new name to chatbot and implement initial skeletal version

jiaxinnns committed:
Echo commands entered by user, and exits when user types bye
'
'

jiaxinnns committed:
Add the ability to store text entered by the user and display them when requested

jiaxinnns committed:
Add the ability to mark tasks as done or undone

jiaxinnns committed:
Add support for tracking three types of tasks (Event, Todo and Deadline)

jiaxinnns committed:
A-TextUiTesting

jiaxinnns committed:
Teach the chatbot to deal with errors such as incorrect inputs entered by the user

jiaxinnns committed:
Add support for deleting tasks from the list

jiaxinnns committed:
Use enums

jiaxinnns committed:
Implement saving of list edits to the hard drive

When the program is run, the existing content in the user's hard drive
will be read. This allows changes to be saved across multiple reruns
of the program.

Editing the completion status, removing a task and adding a task will
automatically update the saved data.

jiaxinnns committed:
Handle situation where data file does not exist yet

jiaxinnns committed:
Merge branch-Level-7 into master

jiaxinnns committed:
Teach the chatbot how to understand dates and times

jiaxinnns committed:
Merge branch-Level-8 into master

jiaxinnns committed:
Refactor the code to extract out closely related code as classes

jiaxinnns committed:
Divide classes into packages

jiaxinnns committed:
Merge branch-A-Packages into master

jiaxinnns committed:
Automate project builds using Gradle

jiaxinnns committed:
Add JUnit Tests

jiaxinnns committed:
Package the app as a JAR file

jiaxinnns committed:
Add JavaDoc comments

jiaxinnns committed:
Follow the coding standard

jiaxinnns committed:
Give users a way to find a task by searching for a keyword

jiaxinnns committed:
Merge branch 'branch-A-JavaDoc'

jiaxinnns committed:
Merge branch-A-CodingStandard into master

jiaxinnns committed:
Merge branch-Level-9 into master

Merge branch-Level-9 into master

jiaxinnns committed:
Add GUI to the chatbot

jiaxinnns committed:
Merge branch 'branch-Level-10'
Merge branch-Level-10 into master

EthanLoww commented on src/main/java/reo/Parser.java:
```diff
@@ -0,0 +1,135 @@
+    private String[] words;
+    private Command command;
+    private TaskList tasks;
+    private Storage storage;
```
Consider leaving a blank line between your fields and methods for better readability

EthanLoww commented on src/main/java/reo/Parser.java:
```diff
@@ -0,0 +1,135 @@
+
+                    String name = firstPart[1];
+                    String from = secondPart[0].trim();
+                    String to = secondPart[1].trim();
+
+                    Event eventToPush = new Event(name, false, to, from);
+                    tasks.addEvent(eventToPush);
+                    storage.writeFile(eventToPush);
+                    return "I've added this event:\n" + eventToPush.toString() + "\n"
+                            + "Now, you have " + tasks.getSize() + " task(s) in your list.\n";
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    return "Please enter a valid task name and to & from dates.";
+                }
+            case DELETE:
+                try {
+                    int index = Integer.valueOf(words[1]) - 1;
+                    // reo.Task toRemove = tasks.get(index);
+                    tasks.deleteTask(index);
+                    storage.removeLine(index + 1);
+                    Task toDelete = tasks.get(index);
+                    return "I've deleted this task:\n" + toDelete.toString() + "\n"
+                            + "Now, you have " + tasks.getSize() + " task(s) in your list.\n";
+                } catch (IndexOutOfBoundsException e) {
+                    return "Please enter a valid task number.\n";
+                }
+            case FIND:
+                try {
+                    String keyword = words[1];
+                    TaskList filtered = new TaskList(tasks.filter(keyword));
+                    return "Here are the matching tasks in your list:\n" + filtered.toString() + "\n";
+                } catch (IndexOutOfBoundsException e) {
+                    return "ERROR: Please enter a valid keyword to search for.";
+                }
+            case BYE:
+                return "Bye!";
+            case UNKNOWN:
+                return "ERROR: Please enter a valid command.\n";
+            default:
+                return "ERROR: Please enter a valid command.\n";
+            }
```
For switch statements, case clauses should not have any indentation.

Method is also too long, consider using helper methods to handle each command.

EthanLoww commented on src/main/java/reo/Storage.java:
```diff
@@ -0,0 +1,177 @@
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+    /**
+     * Returns true if the parameter is 1, and false otherwise.
+     *
+     * @param i The integer used to represent true or false in the saved file.
+     * @return Boolean representation of i.
+     */
+    public static boolean convertIntToBool(int i) {
+        return i == 1;
+    }
```
Instead of doing this conversion, you can do an if check after the switch statement to check whether split[1] is equal to 1 and mark the task accordingly.

EthanLoww commented on src/main/java/reo/Storage.java:
```diff
@@ -0,0 +1,177 @@
+            Scanner s = new Scanner(f);
+            while (s.hasNext()) {
+                String line = s.nextLine();
+                String[] split = line.split("\\|");
```
If you split like this:
String[] split = line.split(" \\| ");
there's no need to trim each of the split parts afterwards

EthanLoww commented on src/main/java/reo/Storage.java:
```diff
@@ -0,0 +1,177 @@
+                String line = s.nextLine();
+                String[] split = line.split("\\|");
+
+                switch (split[0].trim()) {
+                    case "T":
+                        boolean isDoneT = convertIntToBool(Integer.parseInt(split[1].trim()));
+                        String nameT = split[2].trim();
+                        tasks.add(new Todo(nameT, isDoneT));
+                        break;
+                    case "E":
+                        boolean isDoneE = convertIntToBool(Integer.parseInt(split[1].trim()));
+                        String nameE = split[2].trim();
+                        String from = split[3].trim();
+                        String to = split[4].trim();
+                        tasks.add(new Event(nameE, isDoneE, to, from));
+                        break;
+                    case "D":
+                        boolean isDoneD = convertIntToBool(Integer.parseInt(split[1].trim()));
+                        String nameD = split[2].trim();
+                        String deadline = split[3].trim();
+                        tasks.add(new Deadline(nameD, isDoneD, deadline));
+                        break;
+                }
```
Case clauses should not be indented.
Include the default case to check for invalid inputs and handle the error accordingly.
Try to shorten the method by implementing the changes above.

EthanLoww commented on src/main/java/reo/Reo.java:
```diff
@@ -0,0 +1,34 @@
+    public Reo() {
+        TaskList taskList;
+    }
+//    public static void main(String[] args) {
+//        new Reo().run();
+//    }
+//
+//    public void run() {
+//        String currInput;
+//        Scanner scanner = new Scanner(System.in);
+//        currInput = "placeholder";
+//        ui.welcome();
+//        while (!currInput.toLowerCase().equals("bye")) {
+//            currInput = scanner.nextLine().trim();
+//            Parser p = new Parser(currInput, tasklist, ui, storage);
+//            p.parse();
+//        }
+//        scanner.close();
+//    }
```
Delete dead code for better readability

EthanLoww commented on src/main/java/reo/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+    @Override
+    public String toString() {
+        String res = "";
+//        if (tasks.size() == 0) {
+//            return "The list is empty.\n";
+//        }
```
Delete dead code

EthanLoww commented on src/main/java/reo/Storage.java:
```diff
@@ -0,0 +1,177 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+/** Supports the required file operations for reo.Reo. */
```
Javadoc for classes and methods should be detailed and descriptive, while single-line Javadocs are only suitable for class members.

EthanLoww reviewed (commented):
Good job, some code quality violations here and there. Do include javadocs for your public classes and methods, and include a blank line between your fields and methods for better readability.

NicholasT1 commented on src/main/java/reo/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+        this.deadline = stringToDate(deadline);
+    }
+
+    private LocalDate stringToDate(String dateString) {
```
I like how you abstracted this out to ease readability!

NicholasT1 commented on src/main/java/reo/Task.java:
```diff
@@ -0,0 +1,47 @@
+        this.isDone = false;
+    }
+
+    public boolean nameDoesContain(String s) {
```
Maybe the name of the method could sound more like a verb?

NicholasT1 commented on src/main/java/reo/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     * @param taskNumber Index of the task of interest.
+     * @return The String representation at index i.
+     */
+    public String getStringAt(int taskNumber) {
+        return tasks.get(taskNumber).toString();
+    }
```
This method seems abit unnecessary considering Task has a toString() method already

NicholasT1 reviewed (commented):
Overall, your code was neat and easy to read with few style violations. If some longer methods are split up, it would be even easier to read and neater as well.

jiaxinnns committed:
Use Java assertions to document important assumptions that should hold at various points in the code

This helps to detect unexpected situation early.

Let's,
* Check for valid inputs
* Ensure non-null parameters
* Check for valid command formats

jiaxinnns committed:
Critically examine code and refactor to improve the code quality

This makes the code easier for readers to understand.

Lets,
* Avoid 'magic strings' and 'magic literals'
* Split up long methods (more than 30 LoC)

jiaxinnns committed:
Use the Streams feature of Java

jiaxinnns committed:
Merge pull request #2 from jiaxinnns/branch-A-Assertions

Use Java assertions to document important assumptions that should hol…

jiaxinnns committed:
Merge master into branch-A-CodeQuality

jiaxinnns committed:
Merge branch-A-Streams into master

jiaxinnns committed:
Merge pull request #3 from jiaxinnns/branch-A-Streams

Use the Streams feature of Java

jiaxinnns committed:
Merge pull request #1 from jiaxinnns/branch-A-CodeQuality

Critically examine code and refactor to improve the code quality

jiaxinnns committed:
Set up CI

jiaxinnns committed:
Merge remote-tracking branch 'origin/master'
Merge remote master branch into local master

jiaxinnns committed:
Improve code quality in areas missed out previously

jiaxinnns committed:
Add functionality for mass deletion and contact information management

Users may wish to delete multiple contacts / tasks at once, for convenience.

Let's:
* Create new classes and handler functions for contact management
* Modify the delete function to take in multiple indexes for deletion

jiaxinnns committed:
Add JavaDoc comments to new methods

jiaxinnns committed:
Merge branch 'branch-BCD-Extension'
Merge branch-BCD-Extension into master to integrate new features

jiaxinnns committed:
Add a user guide

jiaxinnns committed:
Add a user guide and product screenshot

jiaxinnns committed:
Tweak product website layout

jiaxinnns committed:
Make minor changes to header comments

jiaxinnns committed:
Make Jar file available on Github

jiaxinnns committed:
Update README.md
